pkg/resource/sensitive.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License mapping
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+)
+
+const (
+	errCannotExpandWildcards = "cannot expand wildcards"
+
+	errFmtCannotGetStringForFieldPath  = "cannot not get a string for fieldpath %q"
+	errFmtCannotGetStringsForFieldPath = "cannot get strings matching field fieldpath: %q"
+)
+
+// GetConnectionDetails returns strings matching provided field paths in the
+// input data.
+// See the unit tests for examples.
+func GetConnectionDetails(from map[string]interface{}, at map[string]string) (map[string][]byte, error) {
+	vals := make(map[string][]byte)
+	for tf := range at {
+		vs, err := stringsMatchingFieldPath(from, tf)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtCannotGetStringsForFieldPath, tf)
+		}
+		for k, v := range vs {
+			vals[k] = v
+		}
+	}
+	return vals, nil
+}
+
+func stringsMatchingFieldPath(from map[string]interface{}, path string) (map[string][]byte, error) {
+	paved := fieldpath.Pave(from)
+	segments, err := paved.ExpandWildcards(path)
+	if err != nil {
+		return nil, errors.Wrap(err, errCannotExpandWildcards)
+	}
+
+	res := make(map[string][]byte, len(segments))
+	for _, s := range segments {
+		v, err := paved.GetString(s)
+		if err != nil {
+			return nil, errors.Wrapf(err, errFmtCannotGetStringForFieldPath, s)
+		}
+		res[s] = []byte(v)
+	}
+	return res, nil
+}
+
+// GetSensitiveParameters will collect sensitive information as terraform state
+// attributes by following secret references in the spec.
+func GetSensitiveParameters(ctx context.Context, client SecretClient, from runtime.Object, into map[string]interface{}, mapping map[string]string) error {
+	pv, err := fieldpath.PaveObject(from)
+	if err != nil {
+		return err
+	}
+	xpParams := map[string]interface{}{}
+	if err = pv.GetValueInto("spec.forProvider", &xpParams); err != nil {
```

Comment:
Not sure I get exactly what you mean with _more correct._
In the mapping of the sensitive field, we keep `tfPath`: `xpPath` but we do not append "spec.forProvider" to xpPath path (would be obsolete)
By getting what is under `spec.forProvider`, we kind of normalize the XP parameters and use it. We could also prepend "spec.forProvider" each time we get the fields but I don't see much of a reason to do so.

Reply:
Right now it looks like the following:
```go
map[string]string{"master_password": "masterPasswordSecretRef"}
```
In TFState, `master_password` is at the top level in HCL of the resource definition, whereas `spec.forProvider.masterPasswordSecretRef` would have been the fully qualified path for the Crossplane definition. Both work but the latter is correct in the sense that it's ready to be used in all possible contexts whereas if we have only `masterPasswordSecretRef`, we'd need to know the prefix that's needed to use to actually access the data.

Reply:
Alright, changed like that in the last commit even I don't feel perfect about it since it resulted in some extra string manipulations to trim those prefixes. But sounds good considering it could be used for other contexts.